- Nama: Aghnia Alifya Kamila
- NIM: 24060121140148
- Nama Dataset: Lung Cancer
- Link Dataset:https://archive.ics.uci.edu/dataset/62/lung+cancer

In [ ]:
#Untuk mengimpor pustaka Pandas ke dalam kode Python.
import pandas as pd

In [ ]:
#Variabel url digunakan untuk menyimpan URL tempat dataset breast cancer tersedia.
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/lung-cancer/lung-cancer.data"

#Variabel names berisi daftar nama kolom yang akan digunakan untuk DataFrame yang akan dibaca.
names = ['class', 'Attribute1', 'Attribute2', 'Attribute3', 'Attribute4', 'Attribute5', 'Attribute6', 'Attribute7', 'Attribute8', 'Attribute9']

#Untuk membaca dataset dari URL yang diberikan dan memuatnya ke dalam sebuah DataFrame menggunakan pustaka Pandas.
dataset = pd.read_csv(url, names=names)

In [ ]:
#Melihat dimensi dataset
dataset.shape

(32, 10)

In [ ]:
#Melihat isi dataset
dataset

In [ ]:
# Distribusi Kelas Data
(dataset.groupby('class').size())

In [ ]:
#Ringkasan Statistik
(dataset.describe())

In [ ]:
#Ringkasan Statistik
import matplotlib.pyplot as plt

# Ubah ukuran layout
dataset.plot(kind='box', subplots=True, layout=(3, 4), sharex=False, sharey=False)
plt.show()


In [ ]:
#Visualisasi Data Menggunakan Plot Multivariat
import pandas
from pandas.plotting import scatter_matrix
scatter_matrix(dataset)
plt.show()

In [ ]:
# Visualisasi Data Menggunakan Plot Univariat
dataset.hist()
plt.show()

Trainn-Tes Split

In [ ]:
from sklearn import model_selection

model_selection menyediakan beragam fungsi dan kelas untuk pemisahan dataset menjadi data pelatihan dan data pengujian (train-test split), validasi silang (cross-validation), pemilihan parameter terbaik, dan evaluasi kinerja model, di antara fungsionalitas lainnya

Membuat Validasi Dataset

In [ ]:
array = dataset.values

X = array[:,0:4] #berisikan index ke-0 dan ke-3
Y = array[:,4] #berisikkan index ke-4

validation_size = 0.20

seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

Kode tersebut menggunakan train_test_split() dari modul model_selection di sklearn untuk membagi dataset menjadi data pelatihan (X_train, Y_train) dan data validasi (X_validation, Y_validation). Ini dilakukan dengan mengambil fitur dari kolom indeks 0 hingga 3 untuk X dan label dari kolom indeks 4 untuk Y. Proporsi data validasi diatur menjadi 20% dari keseluruhan dataset dengan nilai seed untuk pengacakan yang konsisten.

In [ ]:
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (25, 4)
X_validation : (7, 4)
Y_train : (25,)
Y_validation : (7,)


Output tersebut menunjukkan dimensi atau ukuran dari setiap variabel setelah dilakukan pemisahan data menggunakan train_test_split().


K-Folds Cross Validation

In [ ]:
seed = 7

scoring = 'accuracy'

Variabel `seed = 7` menetapkan nilai seed untuk pengacakan yang konsisten, sementara variabel `scoring = 'accuracy'` menentukan metrik evaluasi yang digunakan saat melatih model, yaitu akurasi (persentase prediksi yang benar).

Membangun Model

In [ ]:
#import library berisi algoritma klasisfikasi yang akan digunakan
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

# Spot Check Algorithms
models = []

models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))

In [ ]:
# evaluate each model in turn
results = []

names = []

for name, model in models: #Iterasi dilakukan pada setiap model yang disimpan dalam variabel models

  #Setiap model dievaluasi menggunakan cross_val_score() dengan menggunakan metode validasi silang (KFold) yang telah diinisialisasi sebelumnya
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)

  #Hasil evaluasi dari masing-masing lipatan disimpan dalam cv_results.
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)

  #Hasil evaluasi dari setiap model ditambahkan ke dalam list results
  results.append(cv_results)

  #Nama setiap model disimpan dalam list names.
  names.append(name)

  #Dilakukan pencetakan hasil evaluasi masing-masing model beserta rata-rata dan standar deviasi dari performa evaluasi metrik akurasi.
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

KNN: 0.766667 (0.249444)
NB: 0.766667 (0.249444)
SVM: 0.766667 (0.249444)


Hasil output menampilkan rata-rata akurasi dari setiap model klasifikasi (KNN, Naive Bayes, dan SVM) beserta standar deviasinya. Dalam contoh tersebut, rata-rata akurasi dari setiap model adalah 0.766667 dengan standar deviasi sebesar 0.249444. Ini menunjukkan bahwa ketiga model memiliki akurasi yang hampir sama dengan tingkat variasi yang relatif serupa selama proses validasi silang dilakukan.

Memilih Model Terbaik

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [ ]:
#Mendefinisikan model Support Vector Classifier (SVC)
svm = SVC()

#Melatih model SVC dengan menggunakan data pelatihan
svm.fit(X_train, Y_train)

#Memprediksi label untuk data validasi (X_validation) dengan menggunakan model yang telah dilatih.
predictions = svm.predict(X_validation)

#Mencetak tingkat akurasi dari prediksi yang dibuat oleh model pada data validasi, dengan membandingkan prediksi model dengan label sebenarnya dari data validasi.
print(accuracy_score(Y_validation, predictions))

#Mencetak matriks konfusi yang menunjukkan jumlah prediksi yang benar dan salah yang dilakukan oleh model terhadap data validasi untuk setiap kelas.
print(confusion_matrix(Y_validation, predictions))

#Mencetak laporan klasifikasi yang berisi metrik evaluasi seperti presisi, recall, f1-score, dan nilai rata-rata untuk setiap kelas pada prediksi model terhadap data validasi.
print(classification_report(Y_validation, predictions))

0.8571428571428571
[[0 1 0]
 [0 5 0]
 [0 0 1]]
              precision    recall  f1-score   support

           1       0.00      0.00      0.00         1
           2       0.83      1.00      0.91         5
           3       1.00      1.00      1.00         1

    accuracy                           0.86         7
   macro avg       0.61      0.67      0.64         7
weighted avg       0.74      0.86      0.79         7



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Dalam keseluruhan, model memiliki tingkat akurasi yang relatif baik. Namun, perlu diperhatikan bahwa karena kelas 1 memiliki precision dan recall yang rendah, model cenderung sulit untuk memprediksi kelas ini. Evaluasi ini memberikan gambaran tentang kinerja model untuk setiap kelas secara individual, membantu dalam pemahaman terhadap kekuatan dan kelemahan model dalam melakukan prediksi.